runtime-transaction/src/compute_budget_program_id_filter.rs

```diff
@@ -1,37 +1,109 @@
 // static account keys has max
 use {
     agave_transaction_view::static_account_keys_frame::MAX_STATIC_ACCOUNTS_PER_PACKET as FILTER_SIZE,
-    solana_builtins_default_costs::MAYBE_BUILTIN_KEY, solana_sdk::pubkey::Pubkey,
+    solana_builtins_default_costs::{BUILTIN_INSTRUCTION_COSTS, MAYBE_BUILTIN_KEY},
+    solana_sdk::pubkey::Pubkey,
 };
 
+#[derive(Clone, Copy, PartialEq)]
+enum ProgramKind {
+    NotBuiltin,
+    Builtin { is_compute_budget: bool },
+}
+
 pub(crate) struct ComputeBudgetProgramIdFilter {
     // array of slots for all possible static and sanitized program_id_index,
     // each slot indicates if a program_id_index has not been checked (eg, None),
-    // or already checked with result (eg, Some(result)) that can be reused.
-    flags: [Option<bool>; FILTER_SIZE as usize],
+    // or already checked with result (eg, Some(ProgramKind)) that can be reused.
+    program_kind: [Option<ProgramKind>; FILTER_SIZE as usize],
 }
 
 impl ComputeBudgetProgramIdFilter {
     pub(crate) fn new() -> Self {
         ComputeBudgetProgramIdFilter {
-            flags: [None; FILTER_SIZE as usize],
+            program_kind: [None; FILTER_SIZE as usize],
         }
     }
 
     #[inline]
-    pub(crate) fn is_compute_budget_program(&mut self, index: usize, program_id: &Pubkey) -> bool {
-        *self
-            .flags
+    // If instruction at `index` is a builtin, returns Some<is_compute_budget>
+    //   returns None if not a builtin instruction.
+    pub(crate) fn is_builtin_program(&mut self, index: usize, program_id: &Pubkey) -> Option<bool> {
```

Comment:
Actually, how about returning `ProgramKind` here directly and renaming this function to `get_program_kind`? It will make the call-site of this function easier to understand IMO